plugin/web/extensions/sidebar/popups/buildingSearch/components/hooks.ts

```diff
@@ -78,40 +92,188 @@ export default () => {
+    await fetch(`${si.baseURL}/resultsData.csv`)
+      .then(response => response.text())
+      .then(v => {
+        si.resultsData = Papa.parse(v, {
+          header: true,
+          skipEmptyLines: true,
+          fastMode: true,
+        }).data;
+      });
+  }, []);
+
+  useEffect(() => {
+    if (!datasetIndexes?.dataID) return;
+
+    const colorConditions: [string, string][] = [];
+
+    if (highlightAll && resultStyleCondition) {
+      colorConditions.push([resultStyleCondition, "color('red')"]);
+    } else {
+      if (selected) {
+        let selectedConditon = "";
+        selected.forEach(bldg => {
+          if (selectedConditon) selectedConditon += " || ";
+          selectedConditon += "${gml_id} === '" + bldg.gml_id + "'";
+        });
+        if (selectedConditon) {
+          colorConditions.push([selectedConditon, "color('red')"]);
+        }
+      }
+    }
+
+    colorConditions.push(["true", "color()"]);
+
+    const showConditions: [string, string][] = [];
+
+    if (showMatchingOnly) {
+      if (resultStyleCondition) {
+        showConditions.push([resultStyleCondition, "true"]);
+      }
+      showConditions.push(["true", "false"]);
+    } else {
+      showConditions.push(["true", "true"]);
+    }
+
+    const styles = {
+      "3dtiles": {
+        color: {
+          expression: {
+            conditions: colorConditions,
+          },
+        },
+        show: {
+          expression: {
+            conditions: showConditions,
+          },
+        },
+      },
+    };
+
+    postMsg({
+      action: "updateDatasetInScene",
+      payload: {
+        dataID: datasetIndexes.dataID,
+        update: styles,
+      },
+    });
+  }, [selected, resultStyleCondition, datasetIndexes?.dataID, highlightAll, showMatchingOnly]);
 
   const conditionApply = useCallback(() => {
-    // TODO: Search logic here
-    console.log(conditions);
-    const computeResult = TEST_RESULT_DATA;
+    searchResults.current = [];
+    const combinedResults: Result[] = [];
+
+    (async () => {
+      if (searchIndexes.current) {
+        await Promise.all(
+          searchIndexes.current.map(async si => {
+            // get all conditions groups for current search index
+            const threeDTilesId = si.baseURL.split("/").pop() ?? "";
+            const condGroups: DataRowIds[] = [];
+
+            await Promise.all(
+              conditions.map(async cond => {
+                if (cond.values.length > 0) {
+                  const condGroup: DataRowIds = [];
+                  await Promise.all(
+                    cond.values.map(async v => {
+                      await loadDetailData(
+                        `${si.baseURL}/${si.indexRoot.indexes[cond.field].values[v].url}`,
+                      ).then(v => {
+                        condGroup.push(...v.map(item => item.dataRowId));
+                      });
+                    }),
+                  );
+                  condGroups.push(condGroup);
```

Comment:
Does it cause N+1 problem?

Reply:
I believe this is kind of N + 1 problem, but it's requesting independent csv files from server not a database, so maybe cannot optimize unless there's some union search API provided from server end.